Fix Beta calculation for non-daily algorithms

gsalaz98 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
# ⚠️ Warning ⚠️ 

This commit breaks accurate calculations for algorithms that
only make use of `Daily` resolution data. Previously, because
the benchmark was added in Daily resolution in backtesting, any
algorithm that only made use of daily data would have an accurate
calculation for beta and various other statistics.

These changes serve to fix the statistics calculations of non-daily
resolution algorithms, with daily resolution to be revisited at a later
time.

Regression algorithm statistics and a few BacktestNodePacket test statistics have not been updated. They'll be updated once this PR is approved.

Also please note that a few tests should be failing. They are:

* Most regression tests
* QuantConnect.Tests.Common.Packets/BacktestNodePacketTests/ClearsOtherCashAmounts
* QuantConnect.Tests.Common.Packets/BacktestNodePacketTests/InitialCashAmountIsRespected
* QuantConnect.Tests.Common.Packets/BacktestNodePacketTests/JobDatesAreRespected
* QuantConnect.Tests.Engine.Results/BacktestingResultHandlerTests/BasicTemplateDailyAlgorithmSamplesNotMisalignedAbsolute
* QuantConnect.Tests.Engine.Results/BacktestingResultHandlerTests/ResolutionSwitchingAlgorithmSamplesNotMisalignedAbsolute
* QuantConnect.Tests.Engine.Results/BacktestingResultHandlerTests/SamplesNotMisalignedRelative("BasicTemplateDailyAlgorithm")
* QuantConnect.Tests.Engine.Results/BacktestingResultHandlerTests/SamplesNotMisalignedRelative("ResolutionSwitchingAlgorithm")

Any additional failing tests are most likely a new bug.
#### Description
<!--- Describe your changes in detail -->
 Refactors sampling strategy in AlgorithmManager loop

* Refactors AlgorithmManager loop
* Refactors StatisticsBuilder methods and strategy for series alignment
* Move sampling logic to the corresponding IResultHandler
* Changes benchmark resolution to Resolution.Hour
* Modifies IResultHandler to enable external sampling
* Adds BacktestResultHandler unit tests
* Adds ResolutionSwitchingAlgorithm to test misalignment
* Adds support to AlgorithmRunner to store algorithm IResultHandler
* Removed Sample[a-zA-Z]+ methods from IResultHandler definition
* Converted Sample[a-zA-Z]+ methods from public to protected
* Updated inheritors of BaseResultHandler to use new accessibility
modifiers
* Removes useless code in ResolutionSwitchingAlgorithm

#### Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
Closes #3927 
#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Provide accurate statistics to end users. Cleans up AlgorithmManager code.
#### Requires Documentation Change
<!--- Please indicate if these changes will require updates to documentation, and if so, specify what changes are required -->
Encourage users to not use daily resolution if possible

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Unit tests, new unit tests
#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
<!--- The following is a checklist of items that MUST be completed before a PR is accepted -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. <!--- If not applicable, please explain why -->
- [ ] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

<!--- Template inspired by https://www.talater.com/open-source-templates/#/page/99 -->

Martin-Molinero commented on Engine/AlgorithmManager.cs:
```diff
@@ -728,10 +705,8 @@ public void Run(AlgorithmNodePacket job, IAlgorithm algorithm, ISynchronizer syn
-            results.SampleEquity(_previousTime, Math.Round(algorithm.Portfolio.TotalPortfolioValue, 4));
-            SampleBenchmark(results, backtestMode ? _previousTime.Date : _previousTime);
-            SamplePerformance(results, _previousTime.Date, force:true);
+            results.Sample(previousTime, force: true);
```
Why use `previousTime` here? Since we are outside the streaming loop at this stage, wouldn't both be equals? Maybe we can join `time` in the loop with this `previousTime` outside the loop, into a single global `time`?

Martin-Molinero commented on Engine/Results/BacktestingResultHandler.cs:
```diff
@@ -54,6 +55,9 @@ public class BacktestingResultHandler : BaseResultsHandler, IResultHandler
         private readonly HashSet<string> _chartSeriesExceededDataPoints;
 
+        protected decimal _closingPortfolioValue;
+        protected DateTime _previousTime;
```
We could move these into the `BaseResultsHandler`? Maybe to be more precise, could rename to `_previousUtcSampleTime`/`_dailyPortfolioValue`

nit - `using QuantConnect.Lean.Engine.DataFeeds;` not required

Martin-Molinero commented on Engine/Results/LiveTradingResultHandler.cs:
```diff
@@ -1062,6 +1066,43 @@ protected virtual string CreateSafeChartName(string chartName)
+        /// </summary>
+        /// <param name="time">Current time step of the AlgorithmManager loop</param>
+        /// <param name="force">Forces processing of equity and performance if true</param>
+        public virtual void Sample(DateTime time, bool force = false)
```
Seems both live and backtesting implementations are very similar, can we join them and place it into the `BaseResultsHandler`? I think the difference is that Live is sampling the `Benchmark` on each call?

gsalaz98 replied:
Personally, I think it's best that each `Sample` method is implemented separately since they have their own concerns with how they're sampling. Mixing these concerns might make us reliant on specific base class behavior. 

I'll talk to you more about this, would like to know what your thoughts are 👍

gsalaz98 replied:
Update: we looked into addressing this review, but we concluded that it would require additional effort outside the scope of this PR.

Martin-Molinero commented on Common/Statistics/StatisticsBuilder.cs:
```diff
@@ -265,64 +269,45 @@ private static IEnumerable<PeriodRange> GetPeriodRanges(int periodMonths, DateTi
-                Log.Trace("StatisticsBuilder.EnsureSameLength(): Padded Benchmark");
-            }
+            // GroupBy(...) is guaranteed to preserve the order the elements are in.
+            // See http://msdn.microsoft.com/en-us/library/bb534501 for more information.
```
Nice 👍

Martin-Molinero commented on Common/Statistics/StatisticsBuilder.cs:
```diff
@@ -93,23 +93,28 @@ public static class StatisticsBuilder
+            // For good measure and to put any alignment issues to rest, let's resample both collections
+            // to daily resolution just in case.
+            var benchmark = ResampleDaily(ChartPointToDictionary(pointsBenchmark, fromDate, toDate));
+            var performance = ResampleDaily(ChartPointToDictionary(pointsPerformance, fromDate, toDate));
```
Nice 👍

Martin-Molinero commented on Tests/Engine/Results/BacktestingResultHandlerTests.cs:
```diff
@@ -0,0 +1,391 @@
+    [TestFixture]
+    public class BacktestingResultHandlerTests
+    {
+        private Dictionary<string, BacktestingResultHandler> _resultsCache = new Dictionary<string, BacktestingResultHandler>();
```
If we are just consuming the `Charts`, I'd suggest for us to just store them instead, if we need more in the future we can add it too

Martin-Molinero commented on Tests/AlgorithmRunner.cs:
```diff
@@ -44,6 +44,12 @@ namespace QuantConnect.Tests
+        /// <summary>
+        /// Dictionary used to store algorithm <see cref="BacktestingResultHandler"/> instances keyed by the algorithm's <see cref="Language"/>
+        /// </summary>
+        public static Dictionary<Language, Dictionary<string, BacktestingResultHandler>> AlgorithmResults = new Dictionary<Language, Dictionary<string, BacktestingResultHandler>>();
```
I think this isn't required? Since `BacktestingResultHandlerTests` has his own cache? Don't think we can really trust a global cache for results because for example start and end dates could change between executions

gsalaz98 replied:
We use it to extract the BacktestingResultHandler from the AlgorithmRunner and load it into the BacktestingResultHandlerTests. It will overwrite any existing value (if there is one). The only caching we do is localized within the result handler test fixture.

gsalaz98 replied:
I don't think so. We still need a way to get the BacktestingResultHandler out of the AlgorithmRunner call.

Martin-Molinero replied:
Maybe we can create a custom class which holds the `LeanEngineAlgorithmHandlers`, the `AlgorithmManager` and return that instead 🤔

Martin-Molinero commented on Tests/Engine/Results/BacktestingResultHandlerTests.cs:
```diff
@@ -0,0 +1,391 @@
+            var backtestResults = GetResults(algorithm, new DateTime(2013, 10, 7), new DateTime(2013, 10, 11));
+            var benchmarkSeries = backtestResults.Charts["Benchmark"].Series["Benchmark"];
+            var equitySeries = backtestResults.Charts["Strategy Equity"].Series["Equity"];
+            var performanceSeries = backtestResults.Charts["Strategy Equity"].Series["Daily Performance"];
```
I think we are also missing tests that go through the code paths in `StatisticsBuilder`, where part of the changes took place? Maybe simpler unit tests, just for the `StatisticsBuilder`, where for example we can assert it throws if given "misaligned values", it resamples daily

jaredbroad commented:
Closing as tests still not passing. Please open when ready for review.

gsalaz98 commented:
I've gone ahead and changed the existing tests so that they are passing. 

There were a few that couldn't be modified due to the nature of the test, but have been ignored with a message linking to this PR. They are:
* `SamplesNotMisalignedRelative("BasicTemplateDailyAlgorithm")`
* `SamplesNotMisalignedRelative("ResolutionSwitchingAlgorithm")`

I've also addressed Martin's review and updated regression test statistics.

Martin-Molinero commented on Tests/Engine/Results/BacktestingResultHandlerTests.cs:
```diff
@@ -27,19 +28,21 @@ namespace QuantConnect.Tests.Engine.Results
     public class BacktestingResultHandlerTests
     {
-        private Dictionary<string, BacktestingResultHandler> _resultsCache = new Dictionary<string, BacktestingResultHandler>();
+        private object _lock = new object();
```
Why is this lock required?

Martin-Molinero commented on Engine/Results/BaseResultsHandler.cs:
```diff
@@ -106,6 +106,16 @@ public abstract class BaseResultsHandler
+        /// <summary>
+        /// Last time the <see cref="IResultHandler.Sample(DateTime, bool)"/> method was called in UTC
+        /// </summary>
+        protected DateTime _previousUtcSampleTime;
```
nit - since they are now `protected` they should be renamed to `DailyPortfolioValue` and `PreviousUtcSampleTime`

Martin-Molinero commented on Tests/Common/Statistics/StatisticsBuilderTests.cs:
```diff
@@ -0,0 +1,130 @@
+namespace QuantConnect.Tests.Common.Statistics
+{
+    [TestFixture]
+    public class StatisticsBuilderTests
```
Nice 👍

Martin-Molinero commented on Tests/Engine/Results/BacktestingResultHandlerTests.cs:
```diff
@@ -0,0 +1,396 @@
+
+        [TestCase(nameof(BasicTemplateAlgorithm))]
+        [TestCase(nameof(BasicTemplateDailyAlgorithm), Ignore=true, IgnoreReason="Failing. Please see PR for more information: https://github.com/QuantConnect/Lean/pull/4003")]
+        [TestCase(nameof(ResolutionSwitchingAlgorithm), Ignore=true, IgnoreReason="Failing. Please see PR for more information: https://github.com/QuantConnect/Lean/pull/4003")]
```
I think these tests shouldn't be ignored?

Martin-Molinero commented on Tests/Engine/Results/BacktestingResultHandlerTests.cs:
```diff
@@ -0,0 +1,396 @@
+            Assert.AreEqual(0.0, performance.GetAt(1));
+            Assert.AreEqual(Math.Round(-0.01148770, 6), Math.Round(performance.GetAt(2), 6));
+            Assert.AreEqual(Math.Round(0.011161310, 6), Math.Round(performance.GetAt(3), 6));
+            // Unknown value should go here.
```
nit - this comment seems off?

jaredbroad commented on Engine/DataFeeds/UniverseSelection.cs:
```diff
@@ -381,7 +381,7 @@ public bool AddPendingInternalDataFeeds(DateTime utcStart)
                     var dataConfig = _algorithm.SubscriptionManager.SubscriptionDataConfigService.Add(
                         securityBenchmark.Security.Symbol,
-                        _algorithm.LiveMode ? Resolution.Minute : Resolution.Daily,
+                        _algorithm.LiveMode ? Resolution.Minute : Resolution.Hour,
```
127 files changed due to `+2 char; -2 char`....

jaredbroad commented:
Congrats @Martin-Molinero @gsalaz98
